### DISCUSSION TITLE
How to debug C++/WinRT native modules?

### DISCUSSION THREAD

**Discussion by dgellow** [microsoft/react-native-windows#7522] [Q&A] [answered]
Hey there,

I'm currently learning about and experimenting with C++/WinRT native modules. I would be interested to learn from more experienced people what are debugging patterns and strategies. For example what's the way to get logs or traces from the native modules, how do you generate them, how do you visualize them? Can you attach Visual Studio's debugger, or another debugger to a running React Native for Windows application and debug calls to native modules? If yes, could you explain how?

Cheers!

**Comment by dgellow** [comment, marked as answer]
Answering part of my question as that can be useful to others, yes the Visual Studio debugger can be directly attached to the UWP application and breakpoints from the source code work.

Here is a screencast where I set and use a breakpoint then break on an uncaught exception:

![recording (3)](https://user-images.githubusercontent.com/2451004/113474385-02d9ac80-9470-11eb-9f7c-25d492a0310a.gif)

Step by step:
1. Open the `*.sln` file in Visual Studio
2. Run the React Native for Windows via `npx react-native run-windows` (or just `yarn windows`)
3. In Visual Studio, use the menu "Debug > Attach to Process" and select the application from the list. Then click on "Attach"

I still haven't found a way to use logs though.

Edit: I marked this comment as "answer" but feel free to add your comments about debugging strategies, I'm sure I'm not the only one looking for answers regarding this topic :)

**Comment by asklar** [comment]
a native module can use OutputDebugString to write to the debugger output (the Output window in VS when you're debugging).
the JS side can write logs to the web debugger (if web debugging is enabled), via console.log.
Other than that, you should be able to set breakpoints in your native code (F9 on the line you want), and there are ways to debug your JS too but I'm not sure what specific thing you're most interested in so this should be a good starting point.

**Comment by IsuruLakshan97** [comment]
hey can you share the code you ran this fancymath code?

**Reply by dgellow** [reply]
I don't have the project anymore, that was a while ago. But it is based on https://microsoft.github.io/react-native-windows/docs/native-modules.

Are you looking for something specific?
